src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -603,18 +599,47 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 		}
 	}
 
-	/* reject queries which include CommonTableExpr */
+	/* reject queries which include CommonTableExpr which aren't routable */
 	if (queryTree->cteList != NIL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "common table expressions are not supported in distributed "
-							 "modifications",
-							 NULL, NULL);
+		ListCell *cteCell = NULL;
+		DeferredErrorMessage *cteError = NULL;
+
+		foreach(cteCell, queryTree->cteList)
+		{
+			CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+			Query *cteQuery = (Query *) cte->ctequery;
+			bool requiresMasterEvaluation;
+
+			if (cteQuery->commandType != CMD_SELECT)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support non-select common table expressions.",
+									 NULL, NULL);
+			}
+
+			requiresMasterEvaluation = RequiresMasterEvaluation(cteQuery);
+			if (requiresMasterEvaluation)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support common table expressions requiring master evaluation.",
+									 NULL, NULL);
+			}
+
+			/* TODO can we reduce redundancy here by directly calling PlanRouterQuery? */
+			cteError = CreateRouterPlan(cteQuery, cteQuery,
+										plannerRestrictionContext)->planningError;
+			if (cteError != NULL)
+			{
+				return cteError;
+			}
+		}
 	}
 
 	/* extract range table entries */
 	ExtractRangeTableEntryWalker((Node *) queryTree, &rangeTableList);
 
+	/* TODO is this checking CTEs? requiresMasterEvaluation check above shouldn't be necessary if so */
```

Comment:
Per [range_table_walker])(https://github.com/postgres/postgres/blob/master/src/backend/nodes/nodeFuncs.c#L2366), it does not descend into CTEs, while it does descend into subqueries.

I'm wondering whether we should fix this in `ExtractRangeTableEntryWalker` (mostly, how dangerous that would be).

Reply:
`ExtractRangeTableEntryWalker` already descends into CTEs

[https://github.com/citusdata/citus/blob/6741ffd716593f09b22a01dab70f64d7c62922cd/src/backend/distributed/planner/multi_logical_planner.c#L2037] this logic goes to [`query_tree_walker`](https://github.com/postgres/postgres/blob/20e1cc898da6f82b5680973c107e8a96d46caf01/src/backend/nodes/nodeFuncs.c#L2299) without `QTW_IGNORE_CTE_SUBQUERIES`